src/librustc_codegen_ssa/mir/analyze.rs

```diff
@@ -264,7 +148,45 @@ impl<'mir, 'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> Visitor<'tcx>
 
     fn visit_place(&mut self, place: &mir::Place<'tcx>, context: PlaceContext, location: Location) {
         debug!("visit_place(place={:?}, context={:?})", place, context);
-        self.process_place(&place.as_ref(), context, location);
+
+        // Except for `VarDebugInfo`, non-uses do not force locals onto the stack.
+        //
+        // `VarDebugInfo` is handled in `visit_var_debug_info`.
+        if !context.is_use() {
+            return;
+        }
+
+        let mir::Place { local, projection } = *place;
+
+        // Reads from ZSTs do not require memory accesses.
+        if is_consume(context) {
+            let ty = place.ty(self.fx.mir, self.fx.cx.tcx()).ty;
+            let ty = self.fx.monomorphize(&ty);
+            let span = self.fx.mir.local_decls[local].source_info.span;
+            if self.fx.cx.spanned_layout_of(ty, span).is_zst() {
+                return;
+            }
+        }
+
+        assert!(self.place_info == PlaceInfo::default(), "`visit_place` should not recurse");
+        self.visit_projection(local, projection, context, location);
```

Comment:
Can you use a loop instead? I'm hoping we won't need `self.place_info` that way.